File: Frontend/admin/src/api/lpr.ts
/**
 * @file lpr.ts
 * @description LPR 车牌识别通行记录 API
 */

import { httpClient } from '@/utils/alova';

// ==================== 类型定义 ====================

/** 通行方向 */
export type PassDirection = 'entry' | 'exit';

/** 识别结果处理状态 */
export type RecordStatus = 'pending' | 'processed' | 'failed';

/** 车牌识别通行记录 */
export interface PassRecord {
  id: number;
  plate_no: string;
  plate_color: string;
  plate_type: string;
  vehicle_type: string;
  device_id: string;
  device_name: string;
  park_code: string;
  lane_code: string;
  direction: PassDirection;
  pass_time: string;
  image_url: string;
  confidence: number;
  status: RecordStatus;
  related_order_id: string;
  remark: string;
  created_at: string;
  updated_at: string;
}

/** 车辆授权信息 */
export interface VehicleAuthorization {
  is_authorized: boolean;
  auth_type: string;
  driver_name: string;
  driver_phone: string;
  valid_until: string | null;
}

/** 通行记录查询参数 */
export interface PassRecordQueryParams {
  plate_no?: string;
  park_code?: string;
  direction?: PassDirection;
  status?: RecordStatus;
  device_id?: string;
  start_time?: string;
  end_time?: string;
  page?: number;
  page_size?: number;
}

/** 车牌识别回调结果 */
export interface LprCallbackResult {
  success: boolean;
  message: string;
  code: number;
}

/** 车牌识别设备 */
export interface LprDevice {
  id: string;
  name: string;
  sn: string;
  status: 'online' | 'offline';
  park_code: string;
  lane_code?: string;
  direction?: string;
  ip_address?: string;
  last_heartbeat?: string;
  created_at?: string;
}

// ==================== API 函数 ====================

/** 获取设备列表 */
export function listLprDevices(params?: { park_code?: string; status?: string }) {
  return httpClient.get('/v1/lpr/devices', { params });
}

/** 获取设备详情 */
export function getLprDevice(id: string) {
  return httpClient.get(`/v1/lpr/devices/${id}`);
}

/** 查询通行记录列表 */
export function listPassRecords(params?: PassRecordQueryParams) {
  return httpClient.get('/v1/lpr/records', { params });
}

/** 获取通行记录详情 */
export function getPassRecord(id: number) {
  return httpClient.get(`/v1/lpr/records/${id}`);
}

/** 查询车辆授权信息 */
export function getVehicleAuth(plateNo: string, parkCode: string) {
  return httpClient.get('/lpr/vehicle/auth', {
    params: { plate_no: plateNo, park_code: parkCode },
  });
}

/** 获取通行统计 */
export function getPassStats(params?: { park_code?: string; start_date?: string; end_date?: string }) {
  return httpClient.get('/v1/lpr/stats', { params });
}

File: Backend/services/api-gateway/src/grpc_impl/lpr.rs
// ============== LPR 服务 gRPC 调用封装 ==============

impl LprGrpcClient {
    /// 车牌识别回调 — 代理相机推送的识别结果至 lpr-service
    pub async fn lpr_callback(
        &mut self,
        plate_no: String,
        plate_color: String,
        plate_type: String,
        vehicle_type: String,
        pass_time: String,
        image_url: String,
        device_id: String,
        park_code: String,
        direction: String,
        confidence: f64,
        device_name: String,
        lane_code: String,
    ) -> Result<grpc_proto::lpr::LprCallbackResponse, tonic::Status> {
        let request = grpc_proto::lpr::LprCallbackRequest {
            plate_no,
            plate_color,
            plate_type,
            vehicle_type,
            pass_time,
            image_url,
            device_id,
            park_code,
            direction,
            confidence,
            device_name,
            lane_code,
        };
        Ok(self.inner.lpr_callback(request).await?.into_inner())
    }

    /// 查询通行记录列表
    pub async fn list_pass_records(&mut self, plate_no: &str, park_code: &str, direction: &str, status: &str, page: i32, page_size: i32) -> Result<grpc_proto::lpr::ListPassRecordsResponse, tonic::Status> {
        Ok(self.inner.list_pass_records(grpc_proto::lpr::ListPassRecordsRequest {
            plate_no: plate_no.to_string(), park_code: park_code.to_string(),
            direction: direction.to_string(), status: status.to_string(),
            device_id: String::new(), start_time: String::new(), end_time: String::new(),
            page, page_size,
        }).await?.into_inner())
    }

    /// 获取通行记录详情
    pub async fn get_pass_record(&mut self, id: i64) -> Result<grpc_proto::lpr::PassRecordInfo, tonic::Status> {
        Ok(self.inner.get_pass_record(grpc_proto::lpr::GetPassRecordRequest { id }).await?.into_inner())
    }

    /// 获取通行统计
    pub async fn get_pass_stats(&mut self, park_code: &str, start_date: &str, end_date: &str) -> Result<grpc_proto::lpr::PassStatsResponse, tonic::Status> {
        Ok(self.inner.get_pass_stats(grpc_proto::lpr::GetPassStatsRequest {
            park_code: park_code.to_string(), start_date: start_date.to_string(), end_date: end_date.to_string(),
        }).await?.into_inner())
    }

    /// 查询车辆授权
    pub async fn get_vehicle_auth(&mut self, plate_no: &str, park_code: &str) -> Result<grpc_proto::lpr::VehicleAuthInfo, tonic::Status> {
        Ok(self.inner.get_vehicle_auth(grpc_proto::lpr::GetVehicleAuthRequest {
            plate_no: plate_no.to_string(), park_code: park_code.to_string(),
        }).await?.into_inner())
    }
}

